app/models/casebook/concerns/contents.rb

```diff
@@ -0,0 +1,53 @@
+module Casebook::Concerns::Contents
+  extend ActiveSupport::Concern
+
+  def children
+    contents.where(['array_length(casebooks.ordinals, 1) = ?', ordinals.length + 1])
+  end
+
+  def material
+    contents.where.not(material_id: nil)
+  end
+
+  def can_delete?
+    children.count == 0
+  end
+
+  def present?
+    true
+  end
+
+  def reflow_contents
+    reflow_tree tree, ordinals
+  end
+
+  def reflow_tree tree, prefix = []
+    ord = 1
+    tree.each do |element|
+      if element.is_a? Enumerable
+        reflow_tree element, ord == 1 ? prefix : prefix + [ord-1]
+      else
+        element.update_column :ordinals, prefix + [ord]
+        ord += 1
+      end
+    end
+  end
+
+  # Transform a flat sorted array of casebooks into an array tree
+  # [1, 1.1, 1.1.1, 1.2, 2, 2.1, 2.1.1] => [1, [1.1, [1.1.1], 1.2], 2, [2.1, [2.1.1]]]
+  def tree
```

Comment:
Oh, I'd meant to move it from there to here. This is supposed to only be relevant to Books and Chapters (anything that has contents).

It might be more simple to just keep that stuff in Generic, though.